oml/losses/arcface.py

```diff
@@ -0,0 +1,136 @@
+            criterion: Something like CrossEntropyLoss for margin-adjusted OHE
+            label2category: Optional, for label smoothing. If you will not provide it label smoothing will be global and
+                not category-wise
+            label_smoothing: Label smoothing effect strength
+            m: Margin parameter for ArcFace loss. Usually you should use 0.3-0.5 values for it
+            s: Scaling parameter for ArcFace loss. Usually you should use 30-64 values for it
+        """
+        super(ArcFaceLoss, self).__init__()
+
+        assert (
+            label_smoothing is None or 0 < label_smoothing < 1
+        ), f"Choose another label_smoothing parametrization, got {label_smoothing}"
+
+        self.criterion = criterion or nn.CrossEntropyLoss()
+        self.num_classes = num_classes
+        self.label2category = {} if label2category is None else torch.arange(num_classes).apply_(label2category.get)
+        self.label_smoothing = label_smoothing
+        self.weight = nn.Parameter(torch.FloatTensor(num_classes, in_features))
+        nn.init.xavier_uniform_(self.weight)
+        self.rescale = s
+        self.m = m
+        self.cos_m = np.cos(m)
+        self.sin_m = np.sin(m)
+        self.th = -self.cos_m
+        self.mm = self.sin_m * m
+        self.last_logs: Dict[str, float] = {}
+
+    def fc(self, x: torch.Tensor) -> torch.Tensor:
+        return F.linear(F.normalize(x, p=2), F.normalize(self.weight, p=2))
+
+    def smooth_labels(self, y: torch.Tensor) -> torch.Tensor:
+        if isinstance(self.label2category, torch.Tensor):
+            if self.label2category.device.type != self.weight.device.type:
+                self.label2category = self.label2category.to(self.weight.device)
+        return label_smoothing(y, self.num_classes, self.label_smoothing, self.label2category)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        assert torch.all(y < self.num_classes), "You should provide labels between 0 and num_classes - 1."
+
+        cos = self.fc(x)
+
+        self._log_accuracy_on_batch(cos, y)
+
+        sin = torch.sqrt(1.0 - torch.pow(cos, 2))
+
+        cos_w_margin = cos * self.cos_m - sin * self.sin_m
+        cos_w_margin = torch.where(cos > self.th, cos_w_margin, cos - self.mm)
+
+        ohe = F.one_hot(y, self.num_classes)
+        logit = torch.where(ohe.bool(), cos_w_margin, cos) * self.rescale
+
+        if self.label_smoothing:
+            y = self.smooth_labels(y)
+
+        return self.criterion(logit, y)
+
+    @torch.no_grad()
+    def _log_accuracy_on_batch(self, logits: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        accuracy = torch.mean((y == torch.argmax(logits, 1)).to(torch.float32))
+        self.last_logs.update(
+            {
+                ACCURACY_KEY: float(accuracy),
+            }
+        )
+
+
+class ArcFaceLossWithMLP(nn.Module):
+    """
+    Almost the same as ``ArcFaceLoss``, but also has MLP projector before the loss.
```

Comment:
Please, add a comment about motivation: why we may want to have ArcfaceWithMLP while we have ModelWithMLP